src/expr-visitor.cc

```diff
@@ -23,179 +23,333 @@ namespace wabt {
-      break;
-
-    case ExprType::Drop:
-      CHECK_RESULT(delegate_->OnDropExpr(cast<DropExpr>(expr)));
-      break;
-
-    case ExprType::GetGlobal:
-      CHECK_RESULT(delegate_->OnGetGlobalExpr(cast<GetGlobalExpr>(expr)));
-      break;
-
-    case ExprType::GetLocal:
-      CHECK_RESULT(delegate_->OnGetLocalExpr(cast<GetLocalExpr>(expr)));
-      break;
-
-    case ExprType::GrowMemory:
-      CHECK_RESULT(delegate_->OnGrowMemoryExpr(cast<GrowMemoryExpr>(expr)));
-      break;
-
-    case ExprType::If: {
-      auto if_expr = cast<IfExpr>(expr);
-      CHECK_RESULT(delegate_->BeginIfExpr(if_expr));
-      CHECK_RESULT(VisitExprList(if_expr->true_.exprs));
-      CHECK_RESULT(delegate_->AfterIfTrueExpr(if_expr));
-      CHECK_RESULT(VisitExprList(if_expr->false_));
-      CHECK_RESULT(delegate_->EndIfExpr(if_expr));
-      break;
-    }
-
-    case ExprType::Load:
-      CHECK_RESULT(delegate_->OnLoadExpr(cast<LoadExpr>(expr)));
-      break;
+Result ExprVisitor::VisitExpr(Expr* root_expr) {
+  enum class State {
+    Default,
+    Block,
+    IfTrue,
+    IfFalse,
+    Loop,
+    Try,
+    TryCatch,
+  };
+
+  // Use parallel arrays instead of array of structs so we can avoid allocating
+  // unneeded objects. ExprList::iterator has no default constructor, so it
+  // must only be allocated for states that use it.
+  //
+  // NOTE(binji): This can be simplified when the new exception proposal is
+  // implemented, as it only allows for one catch block per try block.
+  std::vector<State> state_stack;
+  std::vector<Expr*> expr_stack;
+  std::vector<ExprList::iterator> expr_iter_stack;
+  std::vector<Index> catch_index_stack;
+
+  auto push_default = [&](Expr* expr) {
+    state_stack.emplace_back(State::Default);
+    expr_stack.emplace_back(expr);
+  };
+
+  auto pop_default = [&]() {
+    state_stack.pop_back();
+    expr_stack.pop_back();
+  };
+
+  auto push_exprlist = [&](State state, Expr* expr, ExprList& expr_list) {
+    state_stack.emplace_back(state);
+    expr_stack.emplace_back(expr);
+    expr_iter_stack.emplace_back(expr_list.begin());
+  };
+
+  auto pop_exprlist = [&]() {
+    state_stack.pop_back();
+    expr_stack.pop_back();
+    expr_iter_stack.pop_back();
+  };
+
+  auto push_catch = [&](Expr* expr, Index catch_index, ExprList& expr_list) {
+    state_stack.emplace_back(State::TryCatch);
+    expr_stack.emplace_back(expr);
+    expr_iter_stack.emplace_back(expr_list.begin());
+    catch_index_stack.emplace_back(catch_index);
+  };
+
+  auto pop_catch = [&]() {
+    state_stack.pop_back();
+    expr_stack.pop_back();
+    expr_iter_stack.pop_back();
+    catch_index_stack.pop_back();
+  };
+
+  push_default(root_expr);
+
+  while (!state_stack.empty()) {
+    State state = state_stack.back();
+    auto* expr = expr_stack.back();
+
+    switch (state) {
+      case State::Default:
+        pop_default();
+
+        switch (expr->type()) {
+          case ExprType::AtomicLoad:
```

Comment:
this is quite a lot of nesting.  Is it worth creating a method for the handling of a single State?